### ISSUE TITLE
[feature or bug?] Multimodal collator ignores `train_on_inputs` / `roles_to_train` / `train_on_eos`; every MM model except Gemma3n trains on full conversation

### ISSUE THREAD

**Issue by thad0ctor** [axolotl-ai-cloud/axolotl#3617]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

I am curious if this is by design or a bug, the consensus online appears to be that training on user prompt/system prompt can be detrimental.

Should the three standard role-masking knobs behave the same way in multimodal training as they do in text-only training, e.g.: 

- `train_on_inputs: false` → loss computed only on assistant turns.
- `roles_to_train: ["assistant"]` → only those roles contribute to the loss.
- `train_on_eos: "turn" | "last" | "all" | "none"` → controls whether EOS/EOT tokens are included in the loss.

Should labels produced by `src/axolotl/prompt_strategies/chat_template.py` (which correctly honors all three options)  be respected by the multimodal collator as well? 

### Current behaviour

In the multimodal training path, all three options are silently ignored. Every token in the sequence — `system` + `user` + role markers + `assistant` + EOS/EOT — contributes to the loss, minus only `pad_token_id` and model-specific image/audio placeholder tokens.

Only `gemma3n`  has an upstream strategy that implements assistant-only masking.

### Evidence from the source (upstream/main @ `7420fd4d`)

1. **`src/axolotl/utils/collators/mm_chat.py`** — only two references to `labels` exist, and there is no branch that preserves dataset-provided labels:

   ```python
   # mm_chat.py:46-57
   batch = self.processing_strategy.processor.apply_chat_template(
       messages,
       add_generation_prompt=False,
       tokenize=True,               # retokenizes from scratch
       return_tensors="pt",
       padding=True,
       return_dict=True,
       chat_template=self.processing_strategy.chat_template,
   )
   # Process the labels
   batch["labels"] = self.processing_strategy.process_labels(batch["input_ids"])
   ```

   Any `labels` produced by preprocessing are discarded: the collator retokenizes `messages` and unconditionally overwrites `batch["labels"]` however regardless, this is moot because VLM documentation states not to enable pre-processing for multimodal models.

2. **`src/axolotl/processing_strategies.py:223-228`** — the base `_mask_non_assistant` is a literal no-op:

   ```python
   def _mask_non_assistant(self, labels: Tensor) -> Tensor:
       """
       Mask non assistant regions to -100.
       To be implemented per subclass.
       """
       return labels
   ```

3. **Subclass audit** — of the 9 `ProcessingStrategy` subclasses upstream, only one overrides `_mask_non_assistant`:

   | Strategy                         | Overrides `_mask_non_assistant`? |
   | -------------------------------- | -------------------------------- |
   | `ProcessingStrategy` (base)      | no-op (`return labels`)          |
   | `Qwen2VLProcessingStrategy`      | No                               |
   | `Qwen3_5ProcessingStrategy`      | No                               |
   | `Gemma3ProcessingStrategy`       | No                               |
   | **`Gemma3nProcessingStrategy`**  | **Yes** — real implementation    |
   | `VoxtralProcessingStrategy`      | No                               |
   | `SmolVLM2ProcessingStrategy`     | No                               |
   | `Mistral3ProcessingStrategy`     | No                               |
   | `InternVLProcessingStrategy`     | No                               |
   | `Glm4vProcessingStrategy`        | No                               |

   Non-`gemma3n` strategies that override `process_labels` only mask media placeholder tokens (image/audio/video), never the user/system/role-marker ranges.

4. **No plumbing for the config knobs.** `grep -rn "train_on_inputs\|roles_to_train\|train_on_eos"` against `src/axolotl/utils/collators/mm_chat.py`, `src/axolotl/processing_strategies.py`, and `src/axolotl/core/builders/` returns **zero hits**. These values are never passed to the collator or any processing strategy.

5. **Instantiation site — `src/axolotl/core/builders/causal.py:522-530`:**

   ```python
   if self.cfg.processor_type and self.processor:
       collator = MultiModalChatDataCollator
       kwargs["processing_strategy"] = get_processing_strategy(
           self.processor,
           training_args.chat_template,
           self.cfg.chat_template,
           image_size=training_args.image_size,
           image_resize_algorithm=training_args.image_resize_algorithm,
       )
   ```

   `train_on_inputs`, `roles_to_train`, `train_on_eos`, `train_on_eot` are not threaded in.

6. **Precedent.** `_mask_non_assistant` was introduced in PR #2852 (`Feat: add gemma3n support`, commit `dfba881e`) as a one-off for Gemma3n, not as a general mechanism. The multimodal rework in PR #2435 predates it and never wired the role-masking options through.

7. **Docs** (`docs.axolotl.ai/docs/multimodal.html`) say only that *"multimodal support is limited and doesn't have full feature parity"* — there is no explicit note that role-masking knobs are inactive for most MM models. Users reading the main YAML docs for `train_on_inputs` / `roles_to_train` reasonably assume they apply.

### Steps to reproduce

1. Pick any non-Gemma3n multimodal recipe (e.g. `examples/qwen2-vl/*`, `examples/gemma3/*`, `examples/llama-4/*`, `examples/internvl/*`, `examples/mistral3/*`, `examples/smolvlm2/*`, `examples/voxtral/*`).
2. Set `train_on_inputs: false` and `roles_to_train: ["assistant"]` in the YAML.
3. Add a small debug hook in `src/axolotl/utils/collators/mm_chat.py::process_rows`, right after `batch["labels"] = ...`:

   ```python
   non_ignored = (batch["labels"] != -100).sum(dim=1)
   total      = (batch["labels"] !=   0 ).shape[1]  # or attention_mask.sum
   print("labels not -100:", non_ignored.tolist(), "/ seq_len:", total)
   # Optionally: decode a slice of kept positions and print the text being trained on.
   ```

4. Run `axolotl train <config>.yaml` on a conversation that has clear `system` / `user` / `assistant` turns.
5. Observe that non-ignored label count ≈ (seq_len − pad − image_placeholder tokens), i.e. roughly the entire sequence. Decoding the kept positions will show the model is being trained on `system` + `user` + role markers + `assistant`, not just `assistant`.
6. Repeat with `chat_template: gemma3n` on a Gemma3n model: the non-ignored count drops to only assistant content + end-of-turn markers, as expected.

### Config yaml

```yaml
# These three options are *silently ignored* by the MM path:
train_on_inputs: false
roles_to_train: ["assistant"]
train_on_eos: turn

whenever the following configs are selected: qwen2_vl, qwen3_5, gemma3, llama3_2_vision, llama4, llava,
   mistral_v7_tekken, pixtral, lfm2vl, internvl, smolvlm, voxtral, mistral3,    
  glm4v, or the fallback ProcessingStrategy (base _mask_non_assistant
   is a no-op, so every non-pad/non-media-placeholder token is in the loss).
```

### Possible solution

Depending on whether this is by design or an actual bug:

The right shape is "derive labels at collation time, honoring the knobs, using the safest available mechanism per model." Three options, in order of cleanliness:

#### (1)  use HF's `return_assistant_tokens_mask=True`

Modern `apply_chat_template` supports `return_assistant_tokens_mask=True`, which uses `{% generation %}` / `{% endgeneration %}` Jinja markers to return a per-token 0/1 assistant mask automatically — correctly through image/audio placeholder expansion. This collapses the problem to:

```python
batch = processor.apply_chat_template(
    messages, ..., tokenize=True, return_dict=True,
    return_assistant_tokens_mask=True,
)
labels = batch["input_ids"].clone()
assistant_mask = torch.tensor(batch["assistant_masks"])
labels[assistant_mask == 0] = -100
# then apply existing media-placeholder masking on top
# then apply train_on_eos / train_on_eot adjustments
```

PR #2787 already fixed `{% generation %}` tag handling in axolotl's Jinja pipeline, so the infrastructure is partly in place. Templates without generation markers would need them added upstream or in the bundled axolotl templates — straightforward change.

#### (2) Fallback: generic turn-marker scanner in base `ProcessingStrategy`

For models whose templates lack `{% generation %}` markers, generalize what `Gemma3nProcessingStrategy` does: take `assistant_start_token` / `assistant_end_token` from the processor or jinja template and scan the retokenized `input_ids`. This collapses Gemma3n's one-off into the default; per-model subclasses only need to handle media tokens.

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main/7420fd4d

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Just a quick answer from the initial question (didn't read full thing), it's because we still don't have full parity for multi-turn training with multimodal collator, so I appreciate the work into this.

**Comment by thad0ctor** [comment]
> Just a quick answer from the initial question (didn't read full thing), it's because we still don't have full parity for multi-turn training with multimodal collator, so I appreciate the work into this.

Thank you , I commited my local patches to add this functionality in https://github.com/axolotl-ai-cloud/axolotl/pull/3628

**Comment by NanoCode012** [comment]
> 1. src/axolotl/utils/collators/mm_chat.py — only two references to labels exist, and there is no branch that preserves dataset-provided labels:

Yes, we skip preprocessing and process on demand for this. Do you plan to preprocess first and re-use it? Curiously, how large is the processed dataset (especially with media)

> src/axolotl/processing_strategies.py:223-228 — the base _mask_non_assistant is a literal no-op:

Yes, that's intended to be filled by each subclass but we did not get to finish work on this. There is some work on gemma3 but never got to generalize it.

> use HF's return_assistant_tokens_mask=True

I saw this before and tried it, however, it's very slow iirc. It does some slow looping internally which would've crushed processing on demand speed. (It's been a while, so can't rmb full details)

> Gemma3nProcessingStrategy does: take assistant_start_token / assistant_end_token from the processor or jinja template and scan the retokenized input_ids.

Ideally, yes, generalize and subclass only if needed. I think if we can fix this part, it would be a nice addition.

**Comment by thad0ctor** [comment]
> > 1. src/axolotl/utils/collators/mm_chat.py — only two references to labels exist, and there is no branch that preserves dataset-provided labels:
> 
> Yes, we skip preprocessing and process on demand for this. Do you plan to preprocess first and re-use it? Curiously, how large is the processed dataset (especially with media)
> 
> > src/axolotl/processing_strategies.py:223-228 — the base _mask_non_assistant is a literal no-op:
> 
> Yes, that's intended to be filled by each subclass but we did not get to finish work on this. There is some work on gemma3 but never got to generalize it.
> 
> > use HF's return_assistant_tokens_mask=True
> 
> I saw this before and tried it, however, it's very slow iirc. It does some slow looping internally which would've crushed processing on demand speed. (It's been a while, so can't rmb full details)
> 
> > Gemma3nProcessingStrategy does: take assistant_start_token / assistant_end_token from the processor or jinja template and scan the retokenized input_ids.
> 
> Ideally, yes, generalize and subclass only if needed. I think if we can fix this part, it would be a nice addition.

1) The approach I kept in https://github.com/axolotl-ai-cloud/axolotl/pull/3625 doesn't pre-process, my multimodal datasets are huge (~40-50GB) so I didn't see the value in updating this and adding more code complexity to the code while clogging my SSD. 

2) Thanks for the input, this is consistent with the PR

3) Noted, I didn't test this method to verify any performance impact and instead ended up going a different route anyways due to {% generation %} / {% endgeneration %}  markers not being present in all the bundled chat templates so I went with a generalized version of the existing Gemma3n scanner as the base instead 

4) Thanks for the input, this is consistent with the PR
